### ISSUE TITLE
ty allows passing different types for constrained TypeVar arguments.

### ISSUE THREAD

**Issue by IDrokin117** [astral-sh/ty#1090]
# Summary
Ty doesn't flag issues when two distinct types are passed to a function that uses the same generic type parameter T for its arguments. This might be allowed for `bound` types (if satisfy), but not for `constraints`. See examples for mypy and pyrefly.
```python
from typing import TypeVar

T = TypeVar("T", int, str) 
def q(s: T, s1: T): # s and s1 must have the same type
    pass
q(1, "str") # not OK
```
```python
from typing import TypeVar

T = TypeVar("T",  bound=int | str) 
def q(s: T, s1: T): # s and s1 might be only subtype of "int | str"
    pass
q(1, "str")  # OK
```
See [Bound versus Constraints](https://www.gaohongnan.com/computer_science/type_theory/05-typevar-bound-constraints.html#id20)
## Playgrounds
- [Ty](https://play.ty.dev/85f7a6c3-e88b-49fd-aa49-d4b85bd0ffe9). No errors
- [Mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=552421be9605bfe11da9e665ae663fef). ```main.py:7: error: Value of type variable "T" of "q" cannot be "object"  [type-var]```
- [Pyrefly](https://pyrefly.org/sandbox/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFCiSyKoZY55QAqihAamRD163KAF4efQSAAUAIm7yANJhQxVAZxggAlPQAmhYFACOszQC4eWgIzXuuy-SiuoCMps31zt1fO0QeV0gA&version=3.12) ```ERROR 7:6-11: Argument `Literal['str']` is not assignable to parameter `s1` with type `int` in function `q` [[bad-argument-type](https://pyrefly.org/en/docs/error-kinds/#bad-argument-type)]```

### Version

0.0.1-alpha.19

**Comment by carljm** [comment]
Thanks for the report! I think at this point this will likely be rolled into #623, but I'll keep it open to make sure it gets fixed.

**Comment by ktbarrett** [comment]
We are seeing the same thing with the following. It's treating the constraint as a union and since the protocol is covariant you can't upcast `Proto[str | bytes]` to `Proto[str] | Proto[bytes]`. But if you treat them as constraints the type of `Proto[AnyString2]` is equivalent to `Proto[str] | Proto[bytes]`.

```python
from typing import TypeVar, Protocol


# This is equivalent to `os.PathLike[AnyStr]`
class Proto[AnyString1: (str, bytes)](Protocol): 
    def spam(self) -> AnyString1: ...


def f[AnyString2: (str, bytes)](path: Proto[AnyString2]) -> None:
    g(path)


def g(path: Proto[str] | Proto[bytes]) -> None:
    del path
```

**Comment by jfuruness** [comment]
This seems to also affect NewType:

```python
  from typing import NewType

  X = NewType("X", int)
  Y = NewType("Y", int)


  def f[T: (X, Y)](a: T, b: T) -> T:
      return a


  f(X(1), X(2))  # ok
  f(Y(1), Y(2))  # ok
  f(X(1), Y(2))  # mypy errors, ty doesn't
  f(Y(1), X(2))  # mypy errors, ty doesn't

  reveal_type(f(X(1), Y(2)))  # ty: `X | Y`
```
